python/packages/autogen-ext/src/autogen_ext/models/llama_cpp/_llama_cpp_completion_client.py

```diff
@@ -0,0 +1,362 @@
+        if finish_reason not in ("stop", "length", "function_calls", "content_filter", "unknown"):
+            finish_reason = "unknown"
+        if thought:
+            create_result = CreateResult(
+                content=content,
+                thought=thought,
+                usage=cast(RequestUsage, response["usage"]),
+                finish_reason=normalize_stop_reason(finish_reason),  # type: ignore
+                cached=False,
+            )
+        else:
+            create_result = CreateResult(
+                content=content,
+                usage=cast(RequestUsage, response["usage"]),
+                finish_reason=normalize_stop_reason(finish_reason),  # type: ignore
+                cached=False,
+            )
+
+        # If we are running in the context of a handler we can get the agent_id
+        try:
+            agent_id = MessageHandlerContext.agent_id()
+        except RuntimeError:
+            agent_id = None
+
+        logger.info(
+            LLMCallEvent(
+                messages=cast(Dict[str, Any], messages),
+                response=create_result.model_dump(),
+                prompt_tokens=response["usage"]["prompt_tokens"],
+                completion_tokens=response["usage"]["completion_tokens"],
+                agent_id=agent_id,
+            )
+        )
+        return create_result
+
+    async def _detect_tool(self, response_text: str, tools: List[Tool]) -> List[FunctionCall]:
+        """
+        Detect if the model is requesting a tool and execute the tool.
+
+        :param response_text: The raw response text from the model.
+        :param tools: A list of available tools.
+        :return: The result of the tool execution or None if no tool is called.
+        """
+        content: List[FunctionCall] = []
+        if not tools:
+            return []
+        for tool in tools:
+            if tool.name.lower() + "(" in response_text.lower():  # Case-insensitive matching
+                logger.debug(f"DEBUG: Detected tool '{tool.name}' in response.")
+                # Extract arguments (if any) from the response
+                func_args = self._extract_tool_arguments(response_text)
+                content.append(
+                    FunctionCall(
+                        id="",
+                        arguments=func_args,
+                        name=normalize_name(tool.name),
+                    )
+                )
+
+        return content
+
+    def _extract_tool_arguments(self, response_text: str) -> str:
+        """
+        Extract tool arguments from the response text.
+
+        :param response_text: The raw response text.
+        :return: A dictionary of extracted arguments.
+        """
+        args_start = response_text.find("{")
+        args_end = response_text.find("}")
+        if args_start != -1 and args_end != -1:
+            return response_text[args_start : args_end + 1]
+
+        return ""
+
+    async def create_stream(
```

Comment:
Should match with

https://github.com/microsoft/autogen/blob/170b8cc8938f1dceeb3db9c0e02a759961adcab3/python/packages/autogen-core/src/autogen_core/models/_model_client.py#L142-L153

Reply:
shouldn't this be async def?

Reply:
Yes the implementation should be async def.